File: backend/app/schemas/maintenance.py
"""Pydantic schemas for maintenance endpoints (Task B4)."""

from datetime import date as date_type
from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class MaintenanceBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    date: date_type | None = None
    odometer: int | None = None
    service_type: str | None = Field(default=None, alias="serviceType")
    category: str | None = None
    cost_cents: int = Field(default=0, alias="costCents")
    currency: str = "USD"
    workshop: str | None = None
    notes: str | None = None


class MaintenanceCreate(MaintenanceBase):
    date: date_type
    service_type: str = Field(alias="serviceType")


class MaintenanceUpdate(MaintenanceBase):
    pass


class MaintenanceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: UUID
    vehicle_id: Annotated[UUID, Field(serialization_alias="vehicleId")]
    date: date_type
    odometer: int | None = None
    service_type: Annotated[str, Field(serialization_alias="serviceType")]
    category: str | None = None
    cost_cents: Annotated[int, Field(serialization_alias="costCents")]
    currency: str
    workshop: str | None = None
    notes: str | None = None
    source: str
    ai_extraction_id: Annotated[UUID | None, Field(serialization_alias="aiExtractionId")]
    created_at: Annotated[datetime, Field(serialization_alias="createdAt")]
    updated_at: Annotated[datetime, Field(serialization_alias="updatedAt")]

File: web/src/server/schemas/maintenance.ts
import { z } from 'zod';

const dateString = z
  .string()
  .regex(/^\d{4}-\d{2}-\d{2}$/, 'date must be YYYY-MM-DD');

/** POST /api/v1/vehicles/{id}/maintenance */
export const createMaintenanceSchema = z.object({
  id: z.string().uuid().optional(),
  date: dateString,
  odometer: z.number().int().nullable().optional(),
  serviceType: z.string().min(1),
  category: z.string().nullable().optional(),
  costCents: z.number().int().default(0),
  currency: z.string().optional(),
  workshop: z.string().nullable().optional(),
  notes: z.string().nullable().optional(),
});

export type CreateMaintenance = z.infer<typeof createMaintenanceSchema>;

/** PATCH /api/v1/maintenance/{id} — all fields optional; no currency re-lock. */
export const updateMaintenanceSchema = z.object({
  date: dateString.nullable().optional(),
  odometer: z.number().int().nullable().optional(),
  serviceType: z.string().min(1).nullable().optional(),
  category: z.string().nullable().optional(),
  costCents: z.number().int().optional(),
  currency: z.string().nullable().optional(),
  workshop: z.string().nullable().optional(),
  notes: z.string().nullable().optional(),
});

export type UpdateMaintenance = z.infer<typeof updateMaintenanceSchema>;

/** Wire shape for MaintenanceRecord responses (Doc 3). */
export const maintenanceResponseSchema = z.object({
  id: z.string().uuid(),
  vehicleId: z.string().uuid(),
  date: z.string(),
  odometer: z.number().int().nullable(),
  serviceType: z.string(),
  category: z.string().nullable(),
  costCents: z.number().int(),
  currency: z.string(),
  workshop: z.string().nullable(),
  notes: z.string().nullable(),
  source: z.string(),
  aiExtractionId: z.string().uuid().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
});

export type MaintenanceResponse = z.infer<typeof maintenanceResponseSchema>;
